### PR TITLE
Add prefer-class-properties rule

### PR DESCRIPTION
**Pull request by lemonmade**
One more rule! Also updated some docs for all the new rules.

cc/ @GoodForOneFare

### CODE REVIEW

File: packages/eslint-plugin-shopify/docs/rules/prefer-class-properties.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+```js
+class Foo {
+  constructor(baz) {
+    this.baz = baz;
+    this.qux = this.initialize();
+  }
+}
+```
+
+The following are considered wranings when setting the argument to `'never'`:
```
Comment (GoodForOneFare) [inline packages/eslint-plugin-shopify/docs/rules/prefer-class-properties.md]: Typo: `wranings`.

File: packages/eslint-plugin-shopify/docs/rules/prefer-class-properties.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+```js
+class Foo {
+  baz = 123;
+  qux = something();
+}
+```
+
+## When Not To Use It
+
+If you do not wish to enforce the use or disuse of Flow for all files, then you can safely disable this rule.
```
Comment (GoodForOneFare) [inline packages/eslint-plugin-shopify/docs/rules/prefer-class-properties.md]: Copy+paste artifact?

File: packages/eslint-plugin-shopify/lib/config/esnext.js [hunk 3]
```diff
@@ -26,6 +26,9 @@ module.exports = {
   rules: merge(
     require('./rules/ecmascript-6'),
     require('./rules/babel'),
-    {'no-param-reassign': 0} // because of default params
+    {
+      'no-param-reassign': 'off', // because of default params
```
Comment (GoodForOneFare) [inline packages/eslint-plugin-shopify/lib/config/esnext.js]: Sucks that this doesn't give a more granular config :/

File: README.md [hunk 4]
```diff
@@ -1786,9 +1807,36 @@ npm run lint
   }
   ```
 
+- [12.11](#12.11) <a name="12.11"></a> When declaring constant instance members in a constructor, prefer class property syntax.
+
+  ESLint rule: [`prefer-class-properties`](packages/eslint-plugin-shopify/docs/prefer-class-properties.md)
+
+  > Why? This often removes the need for an explicit constructor and clearly indicates "setup" members for a class. It also matches how type declarations for instance properties are done in Flow.
```
Comment (GoodForOneFare) [inline README.md]: Really picky, but `It also matches how...` sounds clunky in my head.  I'd prefer something that puts the important part of the sentence first: `It also matches Flow's type declaration and instance property syntax`.

I'm okay with this staying as is, though.

File: packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    return true;
+  }
+
+  function checkConstructorThisAssignment(node) {
+    if (isSimpleLiteral(node.right) && isStaticMemberExpression(node.left)) {
+      context.report({
+        node: node,
+        message: 'BAD!',
```
Comment (GoodForOneFare) [inline packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js]: Is there any chance of eventually making this message more descriptive?

Comment (lemonmade) [inline packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js]: lol, whoops, old commit that didn't get cleaned up :S

File: README.md (hunk 4 above)
Comment (lemonmade) [inline README.md]: You better watch out, keep providing good suggestions and I will make you write all my docs :P

File: packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      });
+    }
+  }
+
+  function checkClassDeclaration(node) {
+    if (applyAlways) {
```
Comment (GoodForOneFare) [inline packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js]: Maybe I just need some coffee beans, but I'd much prefer this to be calling methods with meaningful names.

```
checkClassDeclaration(node)
   if (applyAlways)
    var constructor = findConstructor(node)
   findMemberAssignments(constructor).forEach(... check ...)
  else
    findClassProperties(node).forEach(... BAD ...)
```

Even this has two distinct responsibilities, but I don't have to sift through all of those `===`s to get to the important stuff.

**Comment by GoodForOneFare** [comment]
The usual pedantry from me, but  👍.  The only blocker is a typo and the `When Not To Use It` section.

File: packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js (hunk 6 above)
Comment (lemonmade) [inline packages/eslint-plugin-shopify/lib/rules/prefer-class-properties.js]: Yeah, I was thinking something along these lines too and was too lazy. Good suggestion 👍

**Pull request merged**